internal/pkg/levee/propagation/propagation.go

```diff
@@ -0,0 +1,350 @@
+package propagation
+
+import (
+	"fmt"
+	"go/types"
+	"log"
+	"strings"
+
+	"github.com/google/go-flow-levee/internal/pkg/fieldtags"
+	"github.com/google/go-flow-levee/internal/pkg/sanitizer"
+	"github.com/google/go-flow-levee/internal/pkg/source"
+	"github.com/google/go-flow-levee/internal/pkg/utils"
+	"golang.org/x/tools/go/pointer"
+	"golang.org/x/tools/go/ssa"
+)
+
+type DFSRecord struct {
+	Root         ssa.Node
+	Marked       map[ssa.Node]bool
+	PreOrder     []ssa.Node
+	Sanitizers   []*sanitizer.Sanitizer
+	Config       source.Classifier
+	TaggedFields fieldtags.ResultType
+}
+
+func Dfs(n ssa.Node, conf source.Classifier, taggedFields fieldtags.ResultType) DFSRecord {
+	record := DFSRecord{
+		Root:         n,
+		Marked:       make(map[ssa.Node]bool),
+		PreOrder:     nil,
+		Sanitizers:   nil,
+		Config:       conf,
+		TaggedFields: taggedFields,
+	}
+	var lastBlockVisited *ssa.BasicBlock = nil
+	maxInstrReached := map[*ssa.BasicBlock]int{}
+
+	record.visitReferrers(n, maxInstrReached, lastBlockVisited)
+	record.dfs(n, maxInstrReached, lastBlockVisited, false)
+
+	return record
+}
+
+// dfs performs Depth-First-Search on the def-use graph of the input Source.
+// While traversing the graph we also look for potential sanitizers of this Source.
+// If the Source passes through a sanitizer, dfs does not continue through that Node.
+func (s *DFSRecord) dfs(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock, isReferrer bool) {
```

Comment:
Nit: `s` no longer feels like an appropriate receiver name now that the full name no longer starts with an `s`.

Reply:
Done.